### PR TITLE
Optimizer refactor and cost model additions

### PR DESCRIPTION
**Pull request by GustavoAngulo**
This PR adds a ton of different things before we move over the optimizer to terrier:
* Reorganize the optimizer directory. All stats related files are under `optimizer/stats`, and all cost model files are under `optimizer/cost_model`
* Cost models are now entirely in header files. This is to fit with a theme of portability or "plug-and-play", similar to how indexes are entirely in header files.
* Add in the postgres cost model from 15721 that was never merged in
* Addition of a trivial cost model under `optimizer/cost_model/trivial_cost_model.h`. This can be used for debugging or trivial execution. It does not use any statistics, it can be thought of as heuristic based.
* Fixes a bug involving MurmurHash in `optimizer/stats/stats_util.h`

### CODE REVIEW

File: src/optimizer/postgres_cost_calculator.cpp [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+#include "optimizer/stats/cost.h"
+#include "optimizer/stats/stats_storage.h"
+#include "optimizer/stats/table_stats.h"
+
+#include <cmath>
+
+namespace peloton {
+namespace optimizer {
+
+double PostgresCostCalculator::CalculateCost(
```
[the remark this answers is not in the document]
Comment (GustavoAngulo) [reply src/optimizer/postgres_cost_calculator.cpp]: Postgres does it a bit different as far as I can tell. While we have a cost per tuple, postgres [estimates a cost for hashing a predicate](https://github.com/postgres/postgres/blob/1007b0a126c608b530fd2914825f1e6d133cb649/src/backend/optimizer/path/costsize.c#L3713) to cost [hash joins](https://github.com/postgres/postgres/blob/1007b0a126c608b530fd2914825f1e6d133cb649/src/backend/optimizer/path/costsize.c#L3475-L3476). This is something we could try to implement but would require more complex estimators that should probably be in a separate PR

[the remark this answers is not in the document]
Comment (GustavoAngulo) [reply src/optimizer/postgres_cost_calculator.cpp]: Postgres does it a bit different as far as I can tell. While we have a cost per tuple, postgres [estimates a cost for hashing a predicate](https://github.com/postgres/postgres/blob/1007b0a126c608b530fd2914825f1e6d133cb649/src/backend/optimizer/path/costsize.c#L3713) to cost [hash joins](https://github.com/postgres/postgres/blob/1007b0a126c608b530fd2914825f1e6d133cb649/src/backend/optimizer/path/costsize.c#L3475-L3476). This is something we could try to implement but would require more complex estimators that should probably be in a separate PR

[the remark this answers is not in the document]
Comment (GustavoAngulo) [reply src/optimizer/postgres_cost_calculator.cpp]: Postgres does it a bit different as far as I can tell. While we have a cost per tuple, postgres [estimates a cost for hashing a predicate](https://github.com/postgres/postgres/blob/1007b0a126c608b530fd2914825f1e6d133cb649/src/backend/optimizer/path/costsize.c#L3713) to cost [hash joins](https://github.com/postgres/postgres/blob/1007b0a126c608b530fd2914825f1e6d133cb649/src/backend/optimizer/path/costsize.c#L3475-L3476). This is something we could try to implement but would require more complex estimators that should probably be in a separate PR

**Comment by apavlo** [comment]
@GustavoAngulo I would not worry about getting this to work on Travis since we're not maintaining it.

**Pull request merged**